## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
attrition_df.shape

(1470, 27)

In [3]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [4]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()


Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [5]:
# Create a list of at least 10 column names to use as X data
# Create X_df using your selected columns
X_df = attrition_df[['Education',
                     'Age',
                     'DistanceFromHome',
                     'JobSatisfaction',
                     'OverTime',
                     'StockOptionLevel',
                     'WorkLifeBalance',
                     'YearsAtCompany',
                     'YearsSinceLastPromotion',
                     'NumCompaniesWorked']]
X_df.head()

# Show the data types for X_df
X_df.dtypes


Education                   int64
Age                         int64
DistanceFromHome            int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
dtype: object

In [6]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=1)
X_train.head()

Education  Age  DistanceFromHome  JobSatisfaction OverTime  \
464           3   37                 1                4       No   
512           4   27                 3                4       No   
480           4   30                12                1       No   
680           4   36                 7                4       No   
1023          2   56                 1                1       No   

      StockOptionLevel  WorkLifeBalance  YearsAtCompany  \
464                  0                4               6   
512                  0                3               4   
480                  1                4               1   
680                  0                3               6   
1023                 1                4               3   

      YearsSinceLastPromotion  NumCompaniesWorked  
464                         1                   4  
512                         1                   0  
480                         0                   1  
680                         0                   2  
1023                        1                   2

In [7]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_train["OverTime"].value_counts()

OverTime
No     798
Yes    304
Name: count, dtype: int64

In [8]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
from sklearn.preprocessing import OneHotEncoder
overtime_encoder = OneHotEncoder(sparse_output=False, drop='first', handle_unknown='ignore')
overtime_encoder.fit(np.array(X_train['OverTime']).reshape(-1, 1))
X_train['OverTimeEncoded'] = overtime_encoder.transform(np.array(X_train['OverTime']).reshape(-1, 1))
X_test['OverTimeEncoded'] = overtime_encoder.transform(np.array(X_test['OverTime']).reshape(-1, 1))

X_train = X_train.drop(columns='OverTime')
X_test = X_test.drop(columns='OverTime')

In [ ]:
X_train.dtypes

Education                    int64
Age                          int64
DistanceFromHome             int64
JobSatisfaction              int64
StockOptionLevel             int64
WorkLifeBalance              int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
NumCompaniesWorked           int64
OverTimeEncoded            float64
dtype: object

In [9]:
# Create a StandardScaler
X_scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

X_train_scaled

array([[ 8.08431209e-02, -5.97353422e-04, -1.00933748e+00, ...,
        -3.50759886e-01,  5.05015829e-01, -6.17213400e-01],
       [ 1.07072222e+00, -1.09773647e+00, -7.61501047e-01, ...,
        -3.50759886e-01, -1.09305292e+00, -6.17213400e-01],
       [ 1.07072222e+00, -7.68594737e-01,  3.53762891e-01, ...,
        -6.63492406e-01, -6.93535736e-01, -6.17213400e-01],
       ...,
       [-9.09035982e-01,  3.28544382e-01, -3.89746401e-01, ...,
         1.52563523e+00, -6.93535736e-01,  1.62018517e+00],
       [ 8.08431209e-02,  6.57686118e-01,  8.49435752e-01, ...,
        -3.50759886e-01,  9.04533017e-01,  1.62018517e+00],
       [-9.09035982e-01, -1.42687821e+00,  4.77681106e-01, ...,
        -6.63492406e-01, -6.93535736e-01, -6.17213400e-01]])

In [10]:
# Create a OneHotEncoder for the Department column
# Fit the encoder to the training data
department_encode = OneHotEncoder(sparse_output=False,drop= None,handle_unknown="ignore").fit(np.array(y_train["Department"]).reshape(-1,1))
department_encode.categories_

# Create two new variables by applying the encoder
# to the training and testing data
department_y_train = department_encode.transform(np.array(y_train["Department"]).reshape(-1,1))
department_y_test = department_encode.transform(np.array(y_test["Department"]).reshape(-1,1))

department_y_train


array([[0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       ...,
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.]])

In [11]:
department_y_train.shape

(1102, 3)

In [12]:
# Create a OneHotEncoder for the Attrition column
# Fit the encoder to the training data
attrition_encode = OneHotEncoder(sparse_output=False,drop= None,handle_unknown="ignore").fit(np.array(y_train["Attrition"]).reshape(-1,1))
attrition_encode.categories_

# Create two new variables by applying the encoder
# to the training and testing data
attrition_y_train = attrition_encode.transform(np.array(y_train["Attrition"]).reshape(-1,1))
attrition_y_test = attrition_encode.transform(np.array(y_test["Attrition"]).reshape(-1,1))

attrition_y_train


array([[1., 0.],
       [1., 0.],
       [0., 1.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

In [13]:
attrition_y_train.shape

(1102, 2)

In [14]:
len(X_train.columns)

10

## Create, Compile, and Train the Model

In [15]:
# Find the number of columns in the X training data
num_columns = X_train.shape[1]
num_columns

# Create the input layer
input_layer = layers.Input(shape=(num_columns,), name = "input_features")

# Create at least two shared layers
shared_layer_1 = layers.Dense(units=64, activation="relu", name = "shared_layer_1")(input_layer)
shared_layer_2 = layers.Dense(units=128, activation="relu", name = "shared_layer_2")(shared_layer_1)


In [16]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(units=32, activation="relu", name = "department_hidden")(shared_layer_2)


# Create the output layer
department_output = layers.Dense(department_y_train.shape[1], activation="sigmoid", name = "department_output")(department_hidden)


In [17]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden  = layers.Dense(units=32, activation="sigmoid", name = "attrition_hidden")(shared_layer_2)

# Create the output layer
attrition_output  = layers.Dense(attrition_y_train.shape[1], activation="sigmoid", name = "attrition_output")(attrition_hidden)


In [18]:
# Create the model
#model = Model(inputs=input_layer, outputs=[department_output, attrition_output])
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])
# Compile the model
model.compile(optimizer='adam',
              loss={
                  'department_output': 'categorical_crossentropy',
                  'attrition_output': 'binary_crossentropy'
              },
              metrics={
                  'department_output': 'accuracy',
                  'attrition_output': 'accuracy'
              })

# Summarize the model
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_features            │ (None, 10)             │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared_layer_1 (Dense)    │ (None, 64)             │            704 │ input_features[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared_layer_2 (Dense)    │ (None, 128)            │          8,320 │ shared_layer_1[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_hidden (Dense) │ (None, 32)             │          4,128 │ shared_layer_2[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_hidden (Dense)  │ (None, 32)             │          4,128 │ shared_layer_2[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             99 │ department_hidden[0][… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ attrition_hidden[0][0] │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [19]:
# Train the model
model.fit(
    X_train,
    {
        'department_output': department_y_train,
        'attrition_output': attrition_y_train
    },
    epochs=100,
    batch_size=32,
    validation_split=0.2
)


Epoch 1/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - attrition_output_accuracy: 0.8681 - department_output_accuracy: 0.5555 - loss: 1.7409 - val_attrition_output_accuracy: 0.8281 - val_department_output_accuracy: 0.6335 - val_loss: 1.3075
Epoch 2/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - attrition_output_accuracy: 0.8644 - department_output_accuracy: 0.6805 - loss: 1.1599 - val_attrition_output_accuracy: 0.8281 - val_department_output_accuracy: 0.6290 - val_loss: 1.2977
Epoch 3/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8503 - department_output_accuracy: 0.6434 - loss: 1.2163 - val_attrition_output_accuracy: 0.8281 - val_department_output_accuracy: 0.6335 - val_loss: 1.3030
Epoch 4/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8500 - department_output_accuracy: 0.6580 - loss: 1.1819 - val_attrition_output_accuracy: 0.8281 - val_department_output_accuracy: 0.6244 - val_loss: 1.2903
Epoch 5/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 6

In [ ]:
# Evaluate the model with the testing data
test_results = model.evaluate(
    np.array(X_test),
     {'department_output': np.array(department_y_test),
      'attrition_output': np.array(attrition_y_test)
      }
    )
test_results


12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8064 - department_output_accuracy: 0.5211 - loss: 2.6699 


[2.558657169342041, 0.8260869383811951, 0.5461956262588501]

In [ ]:
# Print the accuracy for both department and attrition
print(f"Department predictions accuracy: {test_results[2]}")
print(f"Attrition predictions accuracy: {test_results[1]}")

Department predictions accuracy: 0.5461956262588501
Attrition predictions accuracy: 0.8260869383811951


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1.
2.
3.

1. The Deparment predictions accuracy score was 54.6% and the Attrition predictions accuracy score was 82.6%.  Based on these results, it appears the Attrition Model is more reliable for predicting employees who are likely to stay with a company compared to the Department Model for predicting the employees that are best suited for the respective deparment due to the low percentage accuracy score.

2. I used the Sigmoid Activation Functions.  The sigmoid activation function is often used in the output layer of neural networks for specific types of problems, particularly binary classification tasks.  Probablity Interpretation:  This sigmoid function outputs values between 0 and 1, which can be interpreted as probabilities.  This makes it particularly useful for binary classification problems where you want to predict teh probability of one class versus another.  Binary Classification:  In binary classification, you typically want to output a single probability value that represents the likelihood of a given input belonging to the positiive class.  The sigmoid function maps the output to this probability range, making it straightforward to apply a threshold (often 0.5) to make the final classification decision.

3.